### ISSUE TITLE
Organize imports in test sources of subprojects does no longer work

### ISSUE THREAD

**Issue by asflierl** [scalameta/metals#2990]
**Describe the bug**

Organize imports in test sources of subprojects does no longer work. Build server is SBT (via BSP).

**To Reproduce**

1. have a (scala) subproject "example" where the test source directory (relative to sbt working dir) is like example/src/test/scala
2. have a source file there with at least one import statement
3. invoke the "organize imports" code action

**Expected behavior**

imports get organized :)

**Actual behavior**

tool tip "no organize import action is available shows"
error in metals log:
```
[Error - 10:45:05 AM] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: scala.meta.internal.metals.ScalafixProvider$ScalafixRunException: SemanticDB not found: example/src/test/scala/Meep.scala
```

**Installation:**

- Operating system: macOS
- Editor: Visual Studio Code
- Metals version: v0.10.5

**Comment by ckipp01** [comment]
Thanks for the report @asflierl. Can you by chance provide an example project with this build structure for us to test? It will make it much easier for us to see what's going on. The "SemanticDB not found" error that you're seeing is often caused when you have compilation issues causing semanticDB not to get produced for that file, or if the file is part of a build target that we don't know about, therefore causing it to not get compiled and therefore not generated.

**Comment by asflierl** [comment]
It only needs the most minimal example of a subproject build. You can use this one: https://github.com/asflierl/organize-imports-minimal-example @ckipp01 

The error can be witnessed in `example/src/test/scala/Meep.scala`

**Comment by asflierl** [comment]
Hmm, I don't know why I thought it only affected sub-projects but it would seem it also doesn't work in the main project. Only the test sources are affected though it seems.

**Comment by asflierl** [comment]
I checked the target folders and the semanticdb files definitely get produced.

**Comment by ckipp01** [comment]
Ahh ok gotcha. So I'm able to reproduce this. Note that it does work fine with Bloop, but when using sbt as a BSP server here it seems not correctly find it. Thanks for the repo @asflierl.

**Comment by tgodzik** [comment]
@ckipp01 it might have something to the targetroot being set wrongly maybe? I add it explicitely in `ScalafixProvider:195`, so you should be able to see what's going  on there.

**Comment by ckipp01** [comment]
So I don't have a fix yet, but I just took a closer look an in this specific situation the issue is indeed the target root is pointing towards the wrong this. For example in the file you list @asflierl, it's the `#example/Test` build target. So the corresponding targetroot for this should be the value of `example / Test / semanticdbTargetRoot`. When checking in sbt this is correctly set to `.../organize-imports-minimal-example/example/target/scala-2.13/test-meta`, but metals is still looking in `../organize-imports-minimal-example/example/target/scala-2.13/meta`. I'm unsure if this is something in our sbt plugin that we aren't doing right or exactly what is going on, but that's the issue.

**Comment by ckipp01** [comment]
Ahhhh actually I think the isssue is that the correct target root is getting appended to the values, and we are grabbing the first one. For example it looks like this:

```
  "-P:semanticdb:targetroot:/Users/ckipp/Documents/scala-workspace/organize-imports-minimal-example/example/target/scala-2.12/meta",
    "-Yrangepos",
    "-P:semanticdb:synthetics:on",
    "-P:semanticdb:failures:warning",
    "-P:semanticdb:targetroot:/Users/ckipp/Documents/scala-workspace/organize-imports-minimal-example/example/target/scala-2.12/test-meta"
```

**Comment by ckipp01** [comment]
Looking at this a bit closer https://github.com/sbt/sbt/blob/develop/main/src/main/scala/sbt/plugins/SemanticdbPlugin.scala, it looks like there is some logic to remove old values and replace them. I'm assuming we'd need to do the same.

We've had a decent amount of issues with getting this all right, or at least I have. I know when we first added this we wanted to fully manage everything to ensure it was all set up correctly for users with various versions. However, I'm becoming more convinced that we're making this way harder on ourselves. Am I correct that instead of all the logic we have in https://github.com/scalameta/metals/blob/main/sbt-metals/src/main/scala/scala/meta/metals/MetalsPlugin.scala we could instead just check the sbt version like we already do and instead just enable the `bspEnabled` setting for the build target and fully let sbt handle it? I know there was hesitations to this originally. @adpi2 do you have maybe any thoughts on this?

**Comment by tgodzik** [comment]
We should at least have a control over the semanticdb version, but maybe everything else can be handled by sbt? Plus maybe add some options that are not being added by sbt itself.

**Comment by adpi2** [comment]
> Am I correct that instead of all the logic we have in https://github.com/scalameta/metals/blob/main/sbt-metals/src/main/scala/scala/meta/metals/MetalsPlugin.scala we could instead just check the sbt version like we already do and instead just enable the bspEnabled setting for the build target and fully let sbt handle it?

It's not that simple but it is not far from it. There are 2 things that sbt don't have by default:
- semanticdb enabled with the latest version
- support of the `debugSession/start` request (that's magically added by the fact that sbt-metals depends on sbt-debug-adapter)

So I think it still makes sense to have a `MetalsPlugin` but it should probably not try to replicate the logic from the `SemanticdbPlugin` because that makes it harder to maintain on both sides (if there is a bug, should we fix it in `MetalsPlugin` or `SemanticdbPlugin` or both?). `SemanticdbPlugin` is quite stable now (at least since 1.5.0).

I suggest something like this:
```scala
override lazy val projectSettings = Def.settings(
  semanticdbVersion := BuildInfo.semanticdbVersion,
  inConfig(Compile)(configSettings),
  inConfig(Test)(configSettings)
)

val configSettings = Def.settings(
  semanticdbEnabled := bspEnabled.value
  semanticdbOptions ++= if (isScala3) Seq() else Seq(
    // Needed for "find references" on implicits and `apply` methods.
    s"-P:semanticdb:synthetics:on",
    // Don't fail compilation in case of Scalameta crash during SemanticDB generation.
    s"-P:semanticdb:failures:warning"
  )
)
```

**Comment by tgodzik** [comment]
I agree with @adpi2 with the exception that we should check whether Scala version is supported and only enable it then. Otherwise if this works I think it's the correct approach.
